File: polyglot/server/py/polyglot_server/autoload_routes.py
import importlib.util
from pathlib import Path
from typing import Any, Dict
from fastapi import FastAPI
from .logger import logger

log = logger.create("autoload_routes", __file__)

def autoload_routes(server: FastAPI, bootstrap: Dict[str, Any]) -> None:
    """
    Autoload route modules from the configured directory.
    Modules must export a 'mount(app)' function.
    """
    if bootstrap.get("routes"):
        routes_dir = Path(bootstrap["routes"])
        log.debug("Loading route modules", {"path": str(routes_dir)})
        if routes_dir.exists():
            module_files = sorted(routes_dir.glob("*.py"))
            log.trace("Found route modules", {"count": len(module_files), "files": [str(f) for f in module_files]})
            for module_path in module_files:
                log.debug("Loading route module", {"module": str(module_path)})
                try:
                    spec = importlib.util.spec_from_file_location(module_path.stem, str(module_path))
                    if spec and spec.loader:
                        module = importlib.util.module_from_spec(spec)
                        spec.loader.exec_module(module)
                        if hasattr(module, "mount"):
                            log.trace("Mounting route module", {"module": module.__name__})
                            module.mount(server)
                        else:
                            log.warn("Route module does not export 'mount' function", {"module": str(module_path)})
                except Exception as e:
                    log.error("Failed to load route module", {"module": str(module_path), "error": str(e)})
                    
            log.info("Route modules loaded", {"count": len(module_files)})
        else:
            log.warn("Routes directory does not exist", {"path": str(routes_dir)})

File: polyglot/server/mjs/src/autoload_routes.mjs
import fs from "fs";
import path from "path";
import { globSync } from "glob";
import logger from "./logger.mjs";

const log = logger.create("autoload_routes", import.meta.url);

/**
 * Autoload route modules from the configured directory.
 * Modules must export a 'mount(server)' function.
 * 
 * @param {import("fastify").FastifyInstance} server 
 * @param {Object} bootstrap 
 */
export async function autoloadRoutes(server, bootstrap) {
    if (bootstrap.routes) {
        log.debug("Loading route modules", { path: bootstrap.routes });
        if (fs.existsSync(bootstrap.routes)) {
            const moduleFiles = globSync(path.join(bootstrap.routes, "*.mjs")).sort();
            log.trace("Found route modules", { count: moduleFiles.length, files: moduleFiles });
            for (const modulePath of moduleFiles) {
                const absolutePath = path.resolve(modulePath);
                log.debug("Loading route module", { module: absolutePath });
                try {
                    const module = await import(absolutePath);
                    if (typeof module.mount === "function") {
                        log.trace("Mounting route module", { module: path.basename(modulePath) });
                        await module.mount(server);
                    } else {
                        log.warn("Route module does not export 'mount' function", { module: absolutePath });
                    }
                } catch (err) {
                    log.error("Failed to load route module", { module: absolutePath, error: err.message });
                }
            }
            log.info("Route modules loaded", { count: moduleFiles.length });
        } else {
            log.warn("Routes directory does not exist", { path: bootstrap.routes });
        }
    }
}
